File: perl5/Perl6.MetaModel/lib/Perl6/MetaClass/Dispatcher.js

if (Perl6 == undefined)       var Perl6           = {};
if (Perl6.MetaClass == undefined) Perl6.MetaClass = function () {};

Perl6.MetaClass.Dispatcher = function (metaclass, order) {
    if (order == undefined)    order = ':canonical';
    if (order == ':canonical') order = ':ascendant';
    var dispatcher;
    if (order == ':ascendant') {
        dispatcher = _make_ascendant_dispatcher(metaclass);
    }
    else {
        throw 'Unsupported dispatch order';
    }
    this.i = dispatcher;
}

Perl6.MetaClass.Dispatcher.prototype.next = function () { return this.i(); }

function _make_iterator (values) {
    var counter = 0;
    return function () {
        if (counter >= values.length) return undefined;
        return values[counter++];
    }
}

function _make_ascendant_dispatcher (metaclass) {
    return _make_iterator(metaclass.MRO());
}

File: perl5/org.perl6.metamodel/src/org/perl6/metamodel/metaclass/Dispatcher.java
package org.perl6.metamodel.metaclass;

import java.util.*;
import org.perl6.metamodel.*;


public class Dispatcher {
	private ArrayList mro;
	private int counter = 0;

	public Dispatcher (MetaClass metaclass) {
		this(metaclass, ":canonical");
	}
	
	public Dispatcher (MetaClass metaclass, String order) {
		try {
			if (order.equals(":canonical"))
				order = ":ascendant";
			if (order.equals(":descendant"))
				_make_descendant_dispatcher(metaclass);
			if (order.equals(":ascendant"))
				_make_ascendant_dispatcher(metaclass);
		} catch (Exception e) {} 	           // This makes me feel dirty.
	}
	
	public boolean hasNext () {
		if (counter < mro.size())
			return true;
		else
			return false;
	}
	
	public MetaClass next () {
		return (MetaClass)mro.get(counter++);
	}
	
	public void remove(){}
	
	// These could be done in the constructor, but this mirrors the other metamodel
	// implementations a little more closely.
	
	private void _make_ascendant_dispatcher (MetaClass metaclass) throws Exception {
		mro = metaclass.MRO();
	}
	
	private void _make_descendant_dispatcher (MetaClass metaclass) throws Exception {
		mro = metaclass.MRO();
		Collections.reverse(mro);
	}
}